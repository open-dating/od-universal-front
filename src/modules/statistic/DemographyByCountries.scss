@import '../../config/variables';

@import "~leaflet/dist/leaflet.css";

.demography-by-countries {
  bottom: 0;
  display: flex;
  flex-direction: column;
  position: absolute;
  top: 0;
  width: 100%;
  height: 100%;
  z-index: 100;

  .demography-by-countries__header {
    height: #{$toolbarHeight};
    display: flex;
    width: 100%;
    position: relative;
  }

  .demography-by-countries__content {
    flex: 1 1 0;
    order: 2;
    position: relative;

    .demography-by-countries__content__map {
      height: 100%;
      width: 100%;

      pre {
        max-height: 200px;
        overflow: auto;
      }
    }
  }

  .fetch-error {
    top: 16px;
    left: 10vw;
    position: absolute;
    width: 80vw;
    z-index: 1000;
  }
}
