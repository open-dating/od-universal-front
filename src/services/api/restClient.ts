import axios, {AxiosError} from 'axios'

import {store} from '../../store'
import {removeUserData} from '../../store/actions/user'
import {i18n} from '../../i18n'

function buildHeaders (token?: string) {
  return {
    authorization: `Bearer ${token}`,
    'accept-language': `${i18n.language}`,
  }
}

function handleError(e: AxiosError, token?: string) {
  if (Number(e.response?.status) === 401 && token) {
    store.dispatch(removeUserData(token) as any)
  }
  return Promise.reject(e)
}

export function post(url: string, data: any, token?: string) {
  return axios.post(url, data, {
    headers: buildHeaders(token),
  }).catch(e => handleError(e, token))
}

export function get(url: string, token?: string, params = {}) {
  return axios.get(url, {
    ...params,
    headers: buildHeaders(token),
  }).catch(e => handleError(e, token))
}

export function patch (url: string, data: any, token?: string) {
  return axios.patch(url, data, {
    headers: buildHeaders(token),
  }).catch(e => handleError(e, token))
}

export function remove(url: string, token?: string) {
  return axios.delete(url, {
    headers: buildHeaders(token),
  }).catch(e => handleError(e, token))
}
