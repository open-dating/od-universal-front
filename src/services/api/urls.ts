import {HOST} from '../../config/config'

function qToQParams(q: any) {
  return Object.keys(q).map(key => `${key}=${q[key]}`).join('&')
}

export const urlsAuth = {
  login: () => `${HOST}/api/v1/auth/login`,
  createUser: () => `${HOST}/api/v1/auth/create-user`,
}

export const urlsUser = {
  create: () => `${HOST}/api/v1/user/create`,
  searchNear: () => `${HOST}/api/v1/user/search-near`,
  profile: (id: number) => `${HOST}/api/v1/user/profile/${id}`,
  myProfile: () => `${HOST}/api/v1/user/my-profile`,
  myProfileSave: () => `${HOST}/api/v1/user/my-profile`,
  settingsSave: () => `${HOST}/api/v1/user/settings`,
  langSave: () => `${HOST}/api/v1/user/lang`,
  saveLocation: () => `${HOST}/api/v1/user/location`,
  saveFcm: () => `${HOST}/api/v1/user/fcm`,
  resetPass: () => `${HOST}/api/v1/auth/auto-change-pass`,
}

export const urlsComplaint = {
  create: () => `${HOST}/api/v1/complaint/create`,
}

export const urlsChoice = {
  like: (id: number) => `${HOST}/api/v1/choice/like/${id}`,
  pass: (id: number) => `${HOST}/api/v1/choice/pass/${id}`,
}

export const urlsIm = {
  sendMessage: () => `${HOST}/api/v1/im/send-message`,
  dialogs: (userId: number, skip: number) => `${HOST}/api/v1/im/dialogs/${userId}?skip=${skip}`,
  messages: (dialogId: number, olderThanId: number) => `${HOST}/api/v1/im/dialog/${dialogId}/messages?olderThanId=${olderThanId}`,
  markAsRead: (dialogId: number) => `${HOST}/api/v1/im/dialog/${dialogId}/mark-as-read`,
  blockDialog: (dialogId: number) => `${HOST}/api/v1/im/dialog/${dialogId}/block`,
}

export const urlsAdmin = {
  users: (q = {}) => `${HOST}/api/v1/admin/users?${qToQParams(q)}`,
  complaints: (q = {}) => `${HOST}/api/v1/admin/complaints?${qToQParams(q)}`,
}

export const urlsPhoto = {
  upload: () => `${HOST}/api/v1/photo/upload`,
}

export const urlsStatistic = {
  user: (relatedToUserId: number) => `${HOST}/api/v1/statistic/user/${relatedToUserId}`,
  publicCountriesBounds: () => `${HOST}/api/v1/statistic/public/countries/bounds`,
  publicDemographyByCountries: () => `${HOST}/api/v1/statistic/public/demography/by-countries`,
}

export const urlsSystem = {
  removeProfile: (id: number)=> `${HOST}/api/v1/system/user/${id}`,
}
