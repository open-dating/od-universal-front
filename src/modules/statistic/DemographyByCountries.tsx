import React, {useEffect, useRef, useState} from 'react'
import * as L from 'leaflet'
import {useSelector} from 'react-redux'

import './DemographyByCountries.scss'
import {StateFetchAnyItem} from '../../interfaces/StateFetchAnyItem'
import {get} from '../../services/api/restClient'
import {urlsStatistic} from '../../services/api/urls'
import {FetchError} from '../../shared-components/FetchError'
import {DemographyDto} from '../../interfaces/DemographyDto'
import {SpinnerPrefetch} from '../../shared-components/SpinnerPrefetch'
import {attachCountryStatPopup, detachCountryStatPopup} from './components/CountryStatPopup'
import {StateApp} from '../../interfaces/StateApp'

export function DemographyByCountries() {
  const userData = useSelector((state: StateApp) => state.user)
  const [stat, setStat] = useState<StateFetchAnyItem<DemographyDto>>({item: null, loading: true, error: null})
  const [bounds, setBounds] = useState()
  const mapElement = useRef<HTMLDivElement>(null)
  const map = useRef<L.Map>()

  const geometry = userData.profile?.location

  useEffect(() => {
    const fetchData = async () => {
      try {
        setStat({item: null, loading: true, error: null})

        const [resp, respBounds] = await Promise.all([
          get(urlsStatistic.publicDemographyByCountries()),
          get(urlsStatistic.publicCountriesBounds()),
        ])

        setStat({item: resp.data, loading: false, error: null})
        setBounds(respBounds.data)
      } catch (e) {
        console.error(e)
        setStat({item: null, loading: false, error: e})
      }
    }

    fetchData()
  }, [])

  useEffect(() => {
    if (!mapElement.current) {
      return
    }
    if (map.current) {
      return
    }

    map.current = L.map(mapElement.current)
      .setView([51.505, -0.09], 3)

    L.tileLayer('http://{s}.tile.openstreetmap.org/{z}/{x}/{y}.png', {
      attribution: 'Map data <a target="_blank" href="http://www.openstreetmap.org">OpenStreetMap.org</a> contributors, ' +
        '<a href="http://creativecommons.org/licenses/by-sa/2.0/">CC-BY-SA</a>',
    }).addTo(map.current)
  }, [])

  useEffect(() => {
    if (!map.current || !bounds) {
      return
    }

    const onEachFeature = (countryCode: string, feature: any, layer: L.GeoJSON) => {
      const wrapper = document.createElement('div')
      const popupLayer = layer.bindPopup(wrapper)

      popupLayer.on('popupopen', () => {
        if (!(stat.item && stat.item[countryCode])) {
          return
        }

        attachCountryStatPopup(wrapper, {
          statItems: stat.item[countryCode].raw,
          feature,
        })
        layer.setStyle({fillColor: '#f5ff64', color: '#000'})
      })

      popupLayer.on('popupclose', () => {
        layer.setStyle({fillColor: '#3388ff', color: '#3388ff'})
      })

      layer.on('remove', () => {
        detachCountryStatPopup(wrapper)
      })
    }

    for (const countryCode in bounds) {
      L.geoJSON(bounds[countryCode], {
        onEachFeature: (f, l: L.GeoJSON) => onEachFeature(countryCode, f, l),
      }).addTo(map.current as unknown as L.Map)
    }

    if (geometry) {
      L.geoJSON(geometry, {
        pointToLayer: function (feature, latlng) {
          return L.marker(latlng, {
            icon: L.divIcon({
              html: 'Me',
              iconSize: [18, 16],
            }),
          })
        },
      }).addTo(map.current as unknown as L.Map)
    }
  }, [stat, bounds, geometry])

  return (
    <div className="demography-by-countries">
      {stat.loading && (<SpinnerPrefetch/>)}
      {stat.error && (<FetchError error={stat.error}/>)}
      <div ref={mapElement} className="demography-by-countries__map"/>
    </div>
  )
}
