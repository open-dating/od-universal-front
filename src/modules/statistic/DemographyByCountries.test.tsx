import React from 'react'
import {Provider} from 'react-redux'
import configureStore from 'redux-mock-store'
import {render} from 'react-dom'
import {act} from 'react-dom/test-utils'
import thunkMiddleware from 'redux-thunk'

import * as restClient from '../../services/api/restClient'
import {DemographyByCountries} from './DemographyByCountries'
import {urlsStatistic} from '../../services/api/urls'

const mockStore = configureStore([thunkMiddleware])

it('Demography by countries, map rendered', async () => {
  jest.spyOn(restClient , 'get').mockImplementation((url) => {
    let data

    if (url === urlsStatistic.publicCountriesBounds()) {
      data = {
        'TW': {
          'type': 'FeatureCollection',
          'licence': 'Data Â© OpenStreetMap contributors, ODbL 1.0. https://osm.org/copyright',
          'features': [],
        },
      }
    } else if (url === urlsStatistic.publicDemographyByCountries()) {
      data = {
        TW: {raw: [], countryName: 'Taiwan'},
      }
    }

    return Promise.resolve({
      headers: {},
      config: {},
      status: 200,
      statusText: '',
      data,
    })
  })

  const store = mockStore({
    user: {
      profile: null,
      jwt: null,
    },
  })

  const container: HTMLElement = document.createElement('div')

  await act(async () => {
    render(<Provider store={store}><DemographyByCountries /></Provider>, container)
  })

})
