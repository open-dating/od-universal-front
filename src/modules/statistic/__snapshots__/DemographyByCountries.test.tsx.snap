// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Demography by countries, map rendered 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <DemographyByCountries>
    <div
      className="demography-by-countries"
    >
      <SpinnerPrefetch>
        <div
          className="spinner-prefetch"
        >
          <div
            className="spinner-prefetch__element"
          >
            <WithStyles(ForwardRef(CircularProgress))>
              <ForwardRef(CircularProgress)
                classes={
                  Object {
                    "circle": "MuiCircularProgress-circle",
                    "circleDisableShrink": "MuiCircularProgress-circleDisableShrink",
                    "circleIndeterminate": "MuiCircularProgress-circleIndeterminate",
                    "circleStatic": "MuiCircularProgress-circleStatic",
                    "colorPrimary": "MuiCircularProgress-colorPrimary",
                    "colorSecondary": "MuiCircularProgress-colorSecondary",
                    "indeterminate": "MuiCircularProgress-indeterminate",
                    "root": "MuiCircularProgress-root",
                    "static": "MuiCircularProgress-static",
                    "svg": "MuiCircularProgress-svg",
                  }
                }
              >
                <div
                  className="MuiCircularProgress-root MuiCircularProgress-colorPrimary MuiCircularProgress-indeterminate"
                  role="progressbar"
                  style={
                    Object {
                      "height": 40,
                      "width": 40,
                    }
                  }
                >
                  <svg
                    className="MuiCircularProgress-svg"
                    viewBox="22 22 44 44"
                  >
                    <circle
                      className="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate"
                      cx={44}
                      cy={44}
                      fill="none"
                      r={20.2}
                      strokeWidth={3.6}
                      style={Object {}}
                    />
                  </svg>
                </div>
              </ForwardRef(CircularProgress)>
            </WithStyles(ForwardRef(CircularProgress))>
          </div>
        </div>
      </SpinnerPrefetch>
      <div
        className="demography-by-countries__header"
      />
      <div
        className="demography-by-countries__content"
      >
        <div
          className="demography-by-countries__content__map"
        />
      </div>
    </div>
  </DemographyByCountries>
</Provider>
`;
